### DISCUSSION TITLE
How can I test debounce with turbine?

### DISCUSSION THREAD

**Discussion by CharAlex** [cashapp/turbine#120] [Q&A]
Hi, I am wondering how can I test the debounce with turbine.
I have the following ViewModel:

```
private val _stepDataChange = MutableSharedFlow<String>()
  val stepDataChange = _stepDataChange.asSharedFlow()

  val performSearchFlow = MutableSharedFlow<String>()

  init {
    viewModelScope.launch(dispatcher) {
      bindPerformSearchIntent()
    }
  }
  
  fun onTextChange(term: String) {
    viewModelScope.launch(dispatcher) {
      performSearchFlow.emit(term)
    }
  }

  private suspend fun bindPerformSearchIntent() {
    performSearchFlow
      .debounce(300)
      .collectLatest { titleText ->
        _stepDataChange.emit(titleText)
      }
  }
```

Here is the test: 

```
@OptIn(ExperimentalCoroutinesApi::class)
  @get: Rule
  var coroutinesTestRule = CoroutineTestRule()

  @ExperimentalCoroutinesApi
  @Test
  fun testDebounce() = runTest {
    val vm = MainViewModel(
      dispatcher = UnconfinedTestDispatcher(TestCoroutineScheduler())
    )

    val flow = vm.stepDataChange.testIn(this)

    vm.onTextChange("Test")

    advanceTimeBy(300)

    assertEquals("Test", flow.awaitItem())

    flow.cancel()
  }
```

where CoroutineTestRule:

```
@ExperimentalCoroutinesApi
class CoroutineTestRule() : TestWatcher() {

  override fun starting(description: Description) {
    super.starting(description)
    Dispatchers.setMain(UnconfinedTestDispatcher(TestCoroutineScheduler()))
  }

  override fun finished(description: Description) {
    super.finished(description)
    Dispatchers.resetMain()
  }
}
```

My test is getting timed out, not sure what am I doing wrong here?
While debugging the breakpoint inside the collectLatest block is never triggered.
If I remove the debounce or set it to 0 the test passes.

Versions:
kotlin: 1.5.31
coroutines: 1.6.1
turbine: 0.8.0

**Comment by mhernand40** [comment]
Looking at your ViewModel, it appears you are already passing in a `CoroutineDispatcher` via the constructor. To rule out any unnecessary factors, your tests don't need to use `CoroutineTestRule`.

As for the test itself, all test dispatchers should be sharing the same scheduler. According to https://developer.android.com/kotlin/coroutines/test#testdispatchers
>**Key Point:** `runTest` runs the test coroutine in a `TestScope`, using a `TestDispatcher`. `TestDispatchers` use a `TestCoroutineScheduler` to control virtual time and schedule new coroutines in a test. All `TestDispatchers` in a test must use the same scheduler instance.

Hence, I'd update
```Kotlin
    val vm = MainViewModel(
      dispatcher = UnconfinedTestDispatcher(TestCoroutineScheduler())
    )
```
to
```Kotlin
    val vm = MainViewModel(
      dispatcher = UnconfinedTestDispatcher(testScheduler)
    )
```
Where `testScheduler` comes from the `this` receiver of `runTest`.

Also, make sure to update to the latest version of Turbine, since there have been some significant changes and fixes lately. Not sure if your tests will suddenly start passing with these changes, but at least you can troubleshoot from there.

**Reply by alexandroschar** [reply]
Thank you for your explanation! Tests are passing now.
